addons/sale_coupon/models/sale_order.py

```diff
@@ -405,14 +414,19 @@ def update_line(order, lines, values):
                             value_found = True
                             # Working on Case 3.
                             lines_to_remove -= line
+                            lines_to_keep += [(0, False, value)]
                             lines_to_remove += update_line(order, line, value)
                             continue
                     # Case 2.
                     if not value_found:
                         lines_to_add += [(0, False, value)]
                 # Case 3.
-                lines_to_remove.unlink()
-                order.write({'order_line': lines_to_add})
+                line_update = []
+                if lines_to_remove:
+                    line_update += [(3, lid, 0) for lid in lines_to_remove.ids]
+                    line_update += lines_to_keep
```

Comment:
Why do you need this line?
`lines_to_keep` could have value that should be deleted and in `lines_to_remove`
```py
# Working on Case 3.
lines_to_remove -= line
lines_to_keep += [(0, False, value)]  # (new line from this PR)
lines_to_remove += update_line(order, line, value)
continue
```

Why keeping a value that might going to be removed?
Either I am tired and this is obvious or this change is a bit strange.

Reply:
I don't know which line exactly you were talking about but there was indeed and issue with line_to_keep which might have some lines that need to be removed in case the value has `0` as qty or price_reduced.
+ Added a testcase

Reply:
I really can't wrap my head around this change. Why the need of the `lines_to_keep` variable (that's the one I was speaking of, and which introduce some weird and hard to read conditions everywhere in this part of the code).

```py
                # Working on Case 3.
                # update_line update the line to the correct value and returns them if they should be unlinked
                update_to_remove = update_line(order, line, value)
                if not update_to_remove:
                    lines_to_keep += [(0, False, value)]
                    lines_to_remove -= line
                continue
        # Working on Case 2.
        if not value_found:
            lines_to_add += [(0, False, value)]
    # Case 3.
    line_update = []
    if lines_to_remove:
        line_update += [(3, line_id, 0) for line_id in lines_to_remove.ids]
        line_update += lines_to_keep
    line_update += lines_to_add
    order.write({'order_line': line_update})
```
=> Why does it need to be so over-complicated? Not saying this is wrong, but an explanation would be welcome.
```py
                # Working on Case 3.
                lines_to_remove -= line
                lines_to_remove += update_line(order, line, value)
                continue
        # Working on Case 2.
        if not value_found:
            lines_to_add += [(0, False, value)]
    # Case 3.
    line_update = [[(3, line_id, 0) for line_id in lines_to_remove.ids]
    line_update += lines_to_add
    order.write({'order_line': line_update})
```